### ISSUE TITLE
Run 2 Instances Of AdGuard Home For Redundancy

### ISSUE THREAD

**Issue by rashidjehangir** [AdguardTeam/AdGuardHome#573]
Hi Guys, is it possible to run 2 instances of AdGuard Home on different PC's on the network for redundancy? I Can use the 2 DNS addresses in the router.

**Comment by ameshkov** [comment]
Well, usually it is possible to configure multiple DNS servers in the router DHCP settings. However, one of them will be the primary server, and it will be used most of the time anyway.

**Comment by iSmigit** [comment]
Think this should be reopened and made a feature request. Routers don't strictly have a primary and secondary DNS in the sense that ones a backup only server, it's just poor terminology that gets used around the place and whether a router favours one DNS server over another is up to implementation and vendor. Generally if multiple DNS servers are configured, they will both be used. Even if one was used more frequently than another, that leaves a percentage of requests going to a different DNS server and if thats not an Adguard server the security measures provided by this product are null and void for those requests.

There's various reasons to want to want to run multiple instances, with the main one being people simply may need to reboot a system from time to time. Either the router will have one DNS server configured which is expected to go down occasionally, or theres multiple and traffic 24/7 will be going to both. Ideally rather than the other DNS server being a public one, it could be another Adguard instances so irrespective of what server is used, the traffic is protected.

Right now thats possible no doubt, but any whitelist, blacklist etc etc will need to be manually applied to two servers. It'd be a great advantage if the web UIs allowed for synchronisation of config which would make managing multiple instances substantially easier and encourage more secure measures. FWIW this is the most upvoted, not implemented, feature over at Pi-Hole so there is demand and I suspect implementing it would win many a customer over.

**Comment by ameshkov** [comment]
@iSmigit well, there's nothing that prevents running multiple instances of AGH as long as they have different configurations / listen ports.

**Comment by iSmigit** [comment]
Thanks for the response. The subject title and my assumption is that the OP and myself would both like to see identically configures instances running on two systems. If one instance was to go down the endpoint or router could direct traffic to the other DNS server, which are both identically configured.

As you note this is already doable, but if you have config that’s not out of the box then you need to manually apply it to both instances. In particular adding filters means managing multiple instances where I think it would be awesome if the application itself could accept the host name of a second instance of the application and have them replicate the system config (filter rules) whenever a change is made to one instance.

Unsure if it’s work well if the DHCP service was enabled, but for pure DNS filtering I think being able to mirror config automatically would be awesome.

**Comment by ameshkov** [comment]
I just think that this is a bit too complicated for the current stage of AGH development.

If we're talking about running two different instances of AGH on two different machines, it *should* be possible to simply `rsync` the config file. The problem is that the sync operation requires restarting the service:
```
stop AGH
rsync AdGuardHome.yaml
start AGH
```

Also, I think we can do it step by step and start with smaller things. For instance, we could start with implementing "reload" and "configtest" operations (like what nginx provides). With these operations it'll be easier to implement settings sync via `rsync`.

**Comment by onedr0p** [comment]
I would like this feature as well, any possibility of opening this issue back up?

**Comment by ameshkov** [comment]
Reopened as a feature request, and issue priority set to "low" for now.

If you upvote this feature request, please also add a comment explaining your use case.

**Comment by onedr0p** [comment]
The important issue to solve would be to allow multiple instances of AdGuardHome to be deployed and synced with each other, for example a Primary, Secondary instance. My use case is to make AdGuardHome to be highly available (HA), thus providing near-zero downtime to DNS requests when my main AdGuardHome instance goes offline. There are many different factors that would allow AdGuardHome to go offline like rebooting for patches or hardware failure.

**Comment by jschwalbe** [comment]
@ameshkov 

> If you upvote this feature request, please also add a comment explaining your use case.

Can you tell me how to upvote it? I would really love this as well. One server goes down, wife says "The wifi is down!! What did you do?" if I have two servers which are synchronized, that buys me some WAF points.

**Comment by onedr0p** [comment]
@jschwalbe and others in this issue. I would check out [Blocky](https://github.com/0xERR0R/blocky). It is completely stateless and can be ran HA with minimal configuration.

**Comment by ameshkov** [comment]
@jschwalbe just ad a 👍 reaction to the issue

**Comment by subdavis** [comment]
## Goal

First, let's clearly establish the purpose of this feature.  It's 100% definitely **not** about load balancing -- if you want load balancing, use HAProxy.  It should be communicated that load balancing is unequivocally out-of-scope for this project.   This issue is about config synchronization, which is a blocker to running multiple instances of AdGuard for High Availability.

I'd like to explore a potential solution which I think would give AdGuard a edge against competitors: webhooks.  I'm interested in contributing here because I think AdGuard's code quality is leagues ahead of PiHole.  I also want to stress that I don't think this issue has been adequately labeled: based on dozens of attempts by users on reddit and pihole forums and hundreds of participants on those threads, a lot of people want this.  AdGuard could be the first to have it.  (Note: Blocky referenced above does not have this feature, it just makes it easier to deploy 2 duplicate instances based on stateless config -- whole different ballgame)

## My Proposal

This is the simplest and most robust solution I could come up with.

Enumerate a set of hookable events, for example:

```go
const (
	DnsConfig       EventType = "dns_config"
	DnsRewrite      EventType = "dns_rewrite"
	DnsSafeBrowsing EventType = "dns_safe_browsing"
	DnsAccess       EventType = "dns_access"
	DnsParental     EventType = "dns_parental"
	DnsSafeSearch   EventType = "dns_safe_search"
	BlockedServices EventType = "blocked_services"
	Dhcp            EventType = "dpcp"
	Stats           EventType = "stats"
	Private         EventType = "private"
	QueryLog        EventType = "query_log"
	Filter          EventType = "filter"
	FilterRule      EventType = "filter_rule"
	I18N            EventType = "i19n"
	Client          EventType = "client"
	Tls             EventType = "tls"
)
```

* Hook into `onConfigModified` to fire a webhook asyncronously when the config is updated, and tell the hook which event caused it.
* Create some new configuration that allows users to specify a webhook destination as the tuple `{ url, auth, []category }`, and fire a webhook at the target(s) when these update.  Have a short timeout, and run the hook in a gofunc or something so it doesn't block if the hook is sluggish.

Then, rather than baking the config sync service into AdGuardHome, we could write another small microservice that waits for webhooks and pulls config from the primary, then pushes it to all secondary nodes.  The microservice could even be very clever and do bi-directional sync if it could diff the changes.  It may actually be better to eventually put this into the core, but webhooks would be a good first step.

**Comment by ameshkov** [comment]
@subdavis let's discuss the proposal here. Otherwise, the discussion will be fragmented.

First of all, thank you for your insights. For some reason, I didn't realize that this feature is that desired. We should definitely find a way to help users configure it.

Regarding webhooks, I generally like the idea of providing them as a feature by itself, it will definitely help people build projects integrated with AdGuard Home, and it will be a great addition to the API. However, I don't think webhooks should be purposefully limited to config changes. The most popular feature requests in this repo are about query logs and statistics, and I suppose webhooks should support them.

About the implementation draft, I think that synchronizing configuration should not require that many different webhook event categories. I kinda understand why you did it -- I guess it is to use existing API methods. However, I think this complicates things, and there is a simpler way.

Alternatively, there can be a single "config changed" event that is fired every time config file changes. The webhook consumer may react to this event, replace the config file on the secondary server, and simply reload AdGuard Home configuration (config reload was implemented in #1302). To make this all easier to implement, we may add "reload" as a separate API method.

Please let me know what you think.

**Comment by subdavis** [comment]
@ameshkov thanks for changing the priority of this issue.

> However, I don't think webhooks should be purposefully limited to config changes.

Sure, expanding the scope to include metrics reporting makes sense.  I haven't looked at those feature requests, but I can.

> About the implementation draft, I think that synchronizing configuration should not require that many different webhook event categories.... there can be a single "config changed" event...

I did this mainly because it allowed a very clean mapping between webhook event name and which API endpoint the webhook handler needs to call to fetch and propagate configuration to "followers".

Here's a example:

* `blocked_services` hook event fires
* webhook handler issues `GET /control/blocked_services/list` against PRIMARY
* webhook handler issues `POST /control/blocked_services/set` against every SECONDARY.

If all the handler gets is `config_changed`, it has to fire *dozens* of API queries on both the primary and secondary servers because it doesn't know what changed.  It has to sync filtering rules and DHCP settings and everything, even though only one thing changed.  IMO, that's unnecessarily heavy and slow.

> The webhook consumer may react to this event, replace the config file on the secondary server

There is no API for fetching a server's entire config, but we could add one. Here's the example I think you're suggesting:

* `config_changed` event fires
* webhook handler issues `GET /control/get_entire_config` against PRIMARY
* webhook handler places config onto disk at secondary's config path.
* webhook handler issues process signal to SECONDARY, causing it to reload config from disk.

* How would the webhook consumer replace the config file on the secondary server unless it's required to be running alongside the secondary?   What if I have 3 or more instances of AdGuard?  What if I'm running these services in a restricted environment like a Synology NAS, which prevents me from sharing application user-space with other processes?
* What if my services are running in Docker? Do I have to bind-mount the config into both AdGuard and the webhook consumer?  I'm also not able to send OS signals to processes running in separate containers without elevating the privilege of the webhook consumer, and this would be very ill-advised.
* What if I don't want to replace the whole file?  maybe my secondary servers should have different passwords.  `tls.server_name` will definitely be different for every server, we shouldn't just overwrite the secondary config with the primary config.  What if I only want to synchronize block and filter lists?

I don't believe mixing filesystem operations, OS signals, and webhooks is a very good practice.

* It's too brittle and rigid - you have to set everything up with perfect permissions and access etc, and it restricts a user's freedom to run each service where/how they want.
* It would require a lot more documentation to explain to users how to configure properly because of all the added requirements and restrictions.

Better to just do the sync purely with REST, I think.

**Comment by ameshkov** [comment]
I'll try to come up with a more detailed description of what I suggest tomorrow.

Meanwhile, a couple of quick thoughts:

1. Exposing methods that work with config file via REST API is not a problem, something like `/control/sync/get_config` and `/control/sync/set_config`.
2. These methods are better to be independent of the other strong-typed structs, maybe even placed to a separate package (`sync`?). Thus we'll guarantee that we won't need to change them regardless of what changes we make to the config structure.
3. We need a separate method that fully reloads AdGuard Home using the new configuration. This one is a bit tricky. I thought we did implement it, but as I see we only have partial reload now. @szolin plz assist.

This would address most of your questions save for one:

> What if I don't want to replace the whole file? maybe my secondary servers should have different passwords.

I suppose this can be handled on the side of the sync micro-service. It's rather easy to exclude specified yaml fields from sync.

**Comment by subdavis** [comment]
> Exposing methods that work with config file via REST API is not a problem

Sure, that seems fine. A bit heavy-handed, perhaps, but definitely less complex for maintainers of both systems.  It may perform badly for users with large filter lists, so that might be worth evaluating.

> These methods are better to be independent of the other strong-typed structs... guarantee that we won't need to change them... 

If you do the whole `/control/sync/get|set` thing, it may make sense to just write the handler in Go and import the Config struct from this package, so you get strong typing for free.

If API consumers use a different language, though, their code is just going to break silently when the schema changes.  This is why I really like Openapi-codegen -- the compiler yells at you when the schema changes and breaks your consumer.

Anyway, thanks for the consideration and the discussion.  I'm happy to help work on this PR, and I'm planning to write the sync handler.  I don't care if that lives in my own personal namespace or this organization.

Cheers.

**Comment by ameshkov** [comment]
> If you do the whole /control/sync/get|set thing, it may make sense to just write the handler in Go and import the Config struct from this package, so you get strong typing for free.

I guess using `configuration` struct is okay, we would need to use it in any case.

> Openapi-codegen -- the compiler yells at you when the schema changes and breaks your consumer.

Well, this is one of the reasons why we're maintaining the openapi spec. We do that manually, though.

**Comment by szolin** [comment]
> We need a separate method that fully reloads AdGuard Home using the new configuration. This one is a bit tricky. I thought we did implement it, but as I see we only have partial reload now. 

We don't have a configuration reload now - currently there's no use for it.
In general, we still need to finish splitting TLS, Web, filter.go modules out from Core (`package home`).  But apart from it, there are no problems to add a Reload() method to each running module.  In fact, TLS module already supports it.

**Comment by ameshkov** [comment]
@szolin got it, thx.

**Comment by yuha0** [comment]
I kind of achieved this and my AdGuard Home setup is running in HA mode. Technically (if I have more servers at home), I can run 100 instances and load balance DNS queries easily, and configuration changes has zero downtime.

However, I did lose some functionalities. I described my setup [here](https://github.com/yuha0/home-infra/tree/master/kubernetes/adguardhome).

Btw, having HA is one of the reason that I made [this feature request](https://github.com/AdguardTeam/AdGuardHome/issues/1964).

**Comment by jeremygaither** [comment]
My use case is simple: I'd like to run two instances in parallel on separate hardware, and somehow synchronize the config and logs. If it can forward logs to a third shared data store, that's a big plus.

**Comment by mattbruman** [comment]
Use case is with some hardware especially some android devices Google or OnePlus has hard coded a secondary DNS as google. Now I have black holed 8.8.8.8 but I get random lag loading sites because the phone sometimes tries to resolve to google anyways if my Adguard DNS sometimes lags a bit. And the only way around this is either switch to static on everything which cant be done on some devices or have a secondary that is different on my router. Setting the same secondary on Unifi does not solve the issue. So now I have two AG running but I have a bunch of filters set for the kids that I sometimes have to disable for their school like youtube. And having to deal with disabling two instances of AG is a hassle, I would rather have the ability to make changes on one reflect on the other, I do have it set in home assistant when i disable one it disables the other but I cant do that with clients when I say disable youtube slider to do the same on the other.

**Comment by abstractvector** [comment]
@mattbruman I’m not sure what firewall / router you’re using, but rather than black-holing 8.8.8.8 you could instead route it back to AdGuard. I’ve done this using the NAT / Port Forwarding feature in OPNsense - all outbound traffic to port 53 gets rerouted to my AdGuard Home server. So several of my devices think they’re getting DNS resolution from 8.8.8.8 but really it’s from AdGuard Home. I haven’t yet needed to set this up for port 853 (DoT) and I don’t know how I’d handle DoH, but it’s working great so far!

**Comment by PSSGCSim** [comment]
@mattbruman 
In my experience putting one IP address two times in the DHCP config solves this issue with Android. If that does not work you, you can simply create secondary virtual IP address on your router and bind AGH to both.

Of course this is a workaround and not a proper HA setup.

Additionally I ICMP reject all requests to common DNS servers (everything including DNS/DoT/DoH).

**Comment by jeremygaither** [comment]
@abstractvector DoT and DoH are very unlikely to work, unless you can install your own CA on all of the devices using AGH, and set up a false certificate for whatever address the device is trying to use (may or may not be Google DNS). Blocking outbound port 853, except from the AdGuard home instance(s), might force the apps/devices to downgrade to normal DNS over port 53 (which can be redirected as mentioned above depending on your router/firewall).

@mattbruman you may be able to use the above along with a second IP address allocated on the AdGuard Home server, so the devices get two unique addresses, yet they point to the same instance. If you are not concerned about downtime, and are only concerned with intermittent lag issues, then you probably don't need HA. But real HA would solve the problem as well, keeping two instances in sync and both could answer queries separately. Perhaps your devices would fall back to the secondary DNS instead of a hard-coded DNS server. Or if you can repro the issue while running wireshark, you may be able to determine what hard-coded resolvers it is attempting to use.

Using rsync to replicate the config file is an interesting workaround (that I may try), but each instance would hold its own metrics/logs. Merging metrics/logs from all instances is also a desirable HA feature. High Availability is probably a high-level epic with multiple tasks, the first being some way for instances to communicate, similar to pfsync from pfsense/opnsense. Something like [ZeroMQ](https://zeromq.org/languages/go/) might be useful, or might be too much, versus a REST-based webhook approach.

**Comment by tomlawesome** [comment]
I would really like the ability to integrate 2+ instances of AdGuard home for redudancy. It's annoying having to run two manually administered setups, each with separate stats. For the block/allow lists at least, it would be good if they could either sync to, or be pulled from a centralised resource, be that a 'primary' instance of AdGuard, or an arbitrary storage location on a server. 

It would be great if the stats could be combined to one pool for analysis (whilst maintaining segregation of data by instance).

**Comment by mzac** [comment]
If anyone is wanting to setup Adguard on a Kubernetes (k8s/k3s) cluster, I got it working on my Raspberry Pi cluster here at home.  I have 3 nodes in my cluster that are running adguard as a daemonset (so they all run one copy).  For the incomming traffic, I am using MetalLB as a load balancer to publish IPs on my network (192.168.0.42 & 192.168.0.43).  In my config below I have also included a service that could be used if you are using keepalived between nodes which is what I was doing before I started using MetalLb.

The only gotcha is that all the nodes will run their own adguard with their own logs, so until we could get centralized logging to work you would have to look at each node to see if it is receiving traffic (maybe with a nodeport).  Another idea is if Adguard supports debugging logging, then kubernetes could spew out the console log to a central log collector.

In order to use MetalLb with k3s, you need to disable the default load balancer that comes with it.  I am using k3sup to provision my cluster with the following config and then deploy MetalLb and Traefik as my ingress controller:

```
export K3S_VERSION="v1.19.5+k3s2"
# Install Master Node - vm1
k3sup install \
    --cluster \
    --ip $MASTER_VM1 \
    --user $SSH_USER \
    --k3s-version $K3S_VERSION \
    --k3s-extra-args '--node-taint key=value:NoSchedule --no-deploy traefik --disable servicelb'
```

If you don't know much about Kubernetes here is a good starting point:
https://blog.alexellis.io/test-drive-k3s-on-raspberry-pi/

Hope this can help someone out there!  Good luck!

```
# --------------------------------------------------------------------------------
# Namespace
---
apiVersion: v1
kind: Namespace
metadata:
  name: adguard

# --------------------------------------------------------------------------------
# Startup Script
---
apiVersion: v1
kind: ConfigMap
metadata:
    name: adguard-init
    namespace: adguard
data:
    adguard-init.sh: |
      #!/bin/sh

      mkdir -p /opt/adguardhome/conf
      cp /tmp/AdGuardHome.yaml /opt/adguardhome/conf
      ls -al /opt/adguardhome/conf/AdGuardHome.yaml

# --------------------------------------------------------------------------------
# Config for Adguard
---
apiVersion: v1
kind: ConfigMap
metadata:
    name: adguard-config
    namespace: adguard
data:
    AdGuardHome.yaml: |
      bind_host: 0.0.0.0
      bind_port: 3000
      users:
      - name: admin
        password: <your password goes here>
      http_proxy: ""
      language: ""
      rlimit_nofile: 0
      debug_pprof: false
      web_session_ttl: 720
      dns:
        bind_host: 0.0.0.0
        port: 53
        statistics_interval: 1
        querylog_enabled: true
        querylog_file_enabled: true
        querylog_interval: 7
        querylog_size_memory: 1000
        anonymize_client_ip: false
        protection_enabled: true
        blocking_mode: default
        blocking_ipv4: ""
        blocking_ipv6: ""
        blocked_response_ttl: 10
        parental_block_host: family-block.dns.adguard.com
        safebrowsing_block_host: standard-block.dns.adguard.com
        ratelimit: 20
        ratelimit_whitelist: []
        refuse_any: true
        upstream_dns:
        - '[/google.com/]tls://8.8.8.8'
        - tls://1.1.1.1
        - tls://1.0.0.1
        upstream_dns_file: ""
        bootstrap_dns:
        - 9.9.9.10
        - 149.112.112.10
        - 2620:fe::10
        - 2620:fe::fe:10
        all_servers: false
        fastest_addr: false
        allowed_clients: []
        disallowed_clients: []
        blocked_hosts:
        - version.bind
        - id.server
        - hostname.bind
        cache_size: 4194304
        cache_ttl_min: 0
        cache_ttl_max: 0
        bogus_nxdomain: []
        aaaa_disabled: false
        enable_dnssec: false
        edns_client_subnet: false
        max_goroutines: 300
        ipset: []
        filtering_enabled: true
        filters_update_interval: 24
        parental_enabled: false
        safesearch_enabled: false
        safebrowsing_enabled: false
        safebrowsing_cache_size: 1048576
        safesearch_cache_size: 1048576
        parental_cache_size: 1048576
        cache_time: 30
        rewrites: []
        blocked_services: []
      tls:
        enabled: false
        server_name: ""
        force_https: false
        port_https: 443
        port_dns_over_tls: 853
        port_dns_over_quic: 784
        allow_unencrypted_doh: false
        strict_sni_check: false
        certificate_chain: ""
        private_key: ""
        certificate_path: ""
        private_key_path: ""
      filters:
      - enabled: true
        url: https://adguardteam.github.io/AdGuardSDNSFilter/Filters/filter.txt
        name: AdGuard DNS filter
        id: 1
      - enabled: false
        url: https://adaway.org/hosts.txt
        name: AdAway Default Blocklist
        id: 2
      - enabled: false
        url: https://www.malwaredomainlist.com/hostslist/hosts.txt
        name: MalwareDomainList.com Hosts List
        id: 4
      whitelist_filters: []
      user_rules: []
      dhcp:
        enabled: false
        interface_name: ""
        dhcpv4:
          gateway_ip: ""
          subnet_mask: ""
          range_start: ""
          range_end: ""
          lease_duration: 86400
          icmp_timeout_msec: 1000
          options: []
        dhcpv6:
          range_start: ""
          lease_duration: 86400
          ra_slaac_only: false
          ra_allow_slaac: false
      clients: []
      log_compress: false
      log_localtime: false
      log_max_backups: 0
      log_max_size: 100
      log_max_age: 3
      log_file: ""
      verbose: false
      schema_version: 7

# --------------------------------------------------------------------------------
# Adguard
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: adguard
  namespace: adguard
  labels:
    app: adguard
spec:
  selector:
    matchLabels:
      app: adguard
  template:
    metadata:
      labels:
        app: adguard
        name: adguard
    spec:
      initContainers:
      - name: adguard-init-script
        image: alpine
        command: ["sh", "/tmp/adguard-init.sh"]
        volumeMounts:
          - name: adguard-config
            mountPath: /opt/adguardhome/conf
          - name: adguard-configmap
            mountPath: /tmp/AdGuardHome.yaml
            subPath: AdGuardHome.yaml
          - name: adguard-init
            mountPath: /tmp/adguard-init.sh
            subPath: adguard-init.sh
      containers:
      - name: adguard
        image: adguard/adguardhome
        imagePullPolicy: IfNotPresent
        livenessProbe:
          httpGet:
            path: /login.html
            port: 3000
            scheme: HTTP
          initialDelaySeconds: 30
          failureThreshold: 5
          timeoutSeconds: 10
        readinessProbe:
          httpGet:
            path: /login.html
            port: 3000
            scheme: HTTP
          initialDelaySeconds: 10
          failureThreshold: 5
          timeoutSeconds: 10
        env:
        - name: TZ
          value: "America/Montreal"
        volumeMounts:
          - name: adguard-config
            mountPath: /opt/adguardhome/conf
      volumes:
        - name: adguard-config
          emptyDir: {}
        - name: adguard-configmap
          configMap:
              name: adguard-config
        - name: adguard-init
          configMap:
              name: adguard-init

# --------------------------------------------------------------------------------
# Service - Adguard DNS - with keepalived
# ---
# apiVersion: v1
# kind: Service
# metadata:
#   name: adguard-dns
#   namespace: adguard
# spec:
#   selector:
#     app: adguard
#   ports:
#   - port: 53
#     targetPort: 53
#     protocol: TCP
#     name: adguard-dns-tcp
#   - port: 53
#     targetPort: 53
#     protocol: UDP
#     name: adguard-dns-udp
#   externalIPs:
#     - 192.168.0.42
#     - 192.168.0.43

# --------------------------------------------------------------------------------
# Service - Adguard DNS - Load Balancer - 192.168.0.42
---
apiVersion: v1
kind: Service
metadata:
  name: adguard-dns-udp-42
  namespace: adguard
  annotations:
    metallb.universe.tf/address-pool: dns
    metallb.universe.tf/allow-shared-ip: dns
spec:
  selector:
    app: adguard
  ports:
  - port: 53
    targetPort: 53
    protocol: UDP
    name: adguard-dns-udp
  type: LoadBalancer
  loadBalancerIP: 192.168.0.42

# --------------------------------------------------------------------------------
# Service - Adguard DNS - Load Balancer - 192.168.0.43
---
apiVersion: v1
kind: Service
metadata:
  name: adguard-dns-udp-43
  namespace: adguard
  annotations:
    metallb.universe.tf/address-pool: dns
    metallb.universe.tf/allow-shared-ip: dns
spec:
  selector:
    app: adguard
  ports:
  - port: 53
    targetPort: 53
    protocol: UDP
    name: adguard-dns-udp
  type: LoadBalancer
  loadBalancerIP: 192.168.0.43

# --------------------------------------------------------------------------------
# Service - Adguard HTTP
---
apiVersion: v1
kind: Service
metadata:
  name: adguard-http
  namespace: adguard
  annotations:
    traefik.ingress.kubernetes.io/service.sticky.cookie: "true"
    traefik.ingress.kubernetes.io/service.sticky.cookie.name: adguard
spec:
  selector:
    app: adguard
  ports:
  - port: 80
    targetPort: 3000
    protocol: TCP
    name: http

# --------------------------------------------------------------------------------
# Ingress
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: adguard-http-ingress
  namespace: adguard
  annotations:
    traefik.ingress.kubernetes.io/router.entrypoints: web
spec:
  rules:
  - host: adguard.lab.local
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: adguard-http
            port:
              name: http
```

**Comment by jeremygaither** [comment]
@mzac Great work, this is similar to what I have wanted to do. Using a DaemonSet is a good idea! Question: are you using the init container because the image doesn't have `/opt/adguardhome/{conf,work}` yet, or for some other reason? It seems like the ConfigMap could be mapped directly to the path used by the app (either default or via CLI option), unless I'm missing something.

As for the query log, I don't know what data structure it uses internally, but a sidecar container could read query log files and push them into Redis or some other data store. Or _something something_ Grafana, since AGH likely wouldn't be easily compatible with Redis or other data stores. A sidecar could also publish Prometheus endpoints, so the perf/query data could get scraped by some other reporting process.

**Comment by mzac** [comment]
> @mzac Great work, this is similar to what I have wanted to do. Using a DaemonSet is a good idea! Question: are you using the init container because the image doesn't have `/opt/adguardhome/{conf,work}` yet, or for some other reason? It seems like the ConfigMap could be mapped directly to the path used by the app (either default or via CLI option), unless I'm missing something.
> 
> As for the query log, I don't know what data structure it uses internally, but a sidecar container could read query log files and push them into Redis or some other data store. Or _something something_ Grafana, since AGH likely wouldn't be easily compatible with Redis or other data stores. A sidecar could also publish Prometheus endpoints, so the perf/query data could get scraped by some other reporting process.

@jeremygaither It has been a few weeks since I got it up and running but I think I needed the init container because adguard was being finicky about how that file wanted to be in the directory.  I think the container was creating the dir and somehow even if I mapped the file as a config map it just wouldn't work which is why I went the init container method to copy it into the correct directory.  I know it is not clean and elegant, but it works!

As for the query log, it looks like it is a plain text/json file in ***/opt/adguardhome/work/data/querylog.json*** so it would just need to be scraped and shipped off somewhere with a sidecar as you suggest.

I would like the idea of sending the logs off into either graylog or elasticsearch directly and then using grafana to create some nice dashboards.

Do you have any ideas if there are any containers out there already that would do that?  (I'm assuming there it I just haven't looked yet)

**Comment by jeremygaither** [comment]
@mzac You might be able to use FluentD to read in the json and pipe them to graylog or elastisearch. I believe I've done this with fluentd piping to elastic and elsewhere for other things (specifically k8s logs and pod logs). I recall existing helm charts for fluentd-elastic that made it fairly easy.

Otherwise, it wouldn't be hard to write a small sidecar specifically for this.

**Comment by onedr0p** [comment]
What about using Grafana's logging tool Loki? It's pretty simple to send logs off to that and have it display in Grafana.

**Comment by tomlawesome** [comment]
@mzac Thank you for posting - this is potentially really useful for me and a great reason to look at setting up a Pi cluster for the first time. I have wanted to integrate traefik somewhere, so it's great to know you have this working in such a way. Following with interest regards logs :) 

Ability to feed into Grafana is ideal.

**Comment by abstractvector** [comment]
I'm not running AdGuard Home in k8s, although a clustered / redundant setup is something I'd like in future. However, I have been successfully ingesting the query log using Telegraf and injecting it into InfluxDB to be rendered by Grafana. I'm no Telegraf expert, but here's the config that's working for me:

```toml
[[inputs.tail]]
  files = ["/opt/AdGuardHome/data/querylog.json"]
  data_format = "json"
  tag_keys = [
    "IP",
    "QH",
    "QT",
    "QC",
    "CP",
    "Upstream",
    "Result_Reason"
  ]

  json_name_key = "query"
  json_time_key = "T"
  json_time_format = "2006-01-02T15:04:05.999999999Z07:00"

[[processors.regex]]
  [[processors.regex.tags]]
    key = "IP"
    result_key = "IP_24"
    pattern = "^(\\d+)\\.(\\d+)\\.(\\d+)\\.(\\d+)$"
    replacement = "${1}.${2}.${3}.x"

[[processors.regex]]
  [[processors.regex.tags]]
    key = "QH"
    result_key = "TLD"
    pattern = "^.*?(?P<tld>[^.]+\\.[^.]+)$"
    replacement = "${tld}"

[[outputs.influxdb]]
  urls = ["http://influxdb.local:8086"]
  database = "adguard"
```

You'll see I've also included a few processors that give me some extra useful stats to chart, including the origin subnet (useful because I use VLANs which map to subnets) and the TLD of the requested domain. It would be easy to run a Telegraf agent in each AdGuard Home pod (mine is actually running in an LXC container on Proxmox rather than k8s) to centralize all your logging.

One thing to note is you'll also want to adjust the `querylog_size_memory` in the `AdGuardHome.yaml` file. This is how many log entries it keeps in memory before flushing to the query log. I think it defaults to 1,000 but I dropped it to 5 to keep the data flowing smoothly to Telegraf.

The end result looks like this:

<img width="1599" alt="Screen Shot 2021-01-14 at 7 26 53 AM" src="https://user-images.githubusercontent.com/726531/104591365-91414d80-563a-11eb-8239-b0b820bcfb85.png">

Hope that helps! @tomlawesome @onedr0p @mzac @jeremygaither

**Comment by mzac** [comment]
@abstractvector That looks great! I'll give telegraf a try, I have used it a few times.  Would you mind sharing the json for your grafana dashboard? (Maybe as a gist?)

**Comment by abstractvector** [comment]
> @abstractvector That looks great! I'll give telegraf a try, I have used it a few times. Would you mind sharing the json for your grafana dashboard? (Maybe as a gist?)

Sure thing, here you go: https://gist.github.com/abstractvector/7de922c17bf3948792f42f71439959e0

There are a few field aliases in there for my DNS upstreams that you'll want to update, and you'll also need the pie chart plugin on Grafana if you don't already.

I'm no Grafana expert, so there may well be better ways of doing some of the things I've done / presenting the data, so feel free to use it as a starting point for building something better!

**Comment by mzac** [comment]
> > @abstractvector That looks great! I'll give telegraf a try, I have used it a few times. Would you mind sharing the json for your grafana dashboard? (Maybe as a gist?)
> 
> Sure thing, here you go: https://gist.github.com/abstractvector/7de922c17bf3948792f42f71439959e0
> 
> There are a few field aliases in there for my DNS upstreams that you'll want to update, and you'll also need the pie chart plugin on Grafana if you don't already.
> 
> I'm no Grafana expert, so there may well be better ways of doing some of the things I've done / presenting the data, so feel free to use it as a starting point for building something better!

Thanks!  I got the telegraf sidecar running and pushing into influx and now the Grafana dashboard is working!  Very awesome!

Now to explore the data a little more...

**Comment by filikun** [comment]
User case is like most here to be able to have two instances of Adguard home syncing settings and stats between the two. Let one be the controller and the other slave of some sort.

**Comment by mzac** [comment]
> User case is like most here to be able to have two instances of Adguard home syncing settings and stats between the two. Let one be the controller and the other slave of some sort.

Sounds just about right.  They could synchronize in the backend with a database (mongo, redis, sql, etc).  That way you don't really need a 'master' per say.  It would be nice if when changing settings on one the others would notice it right away without having to reload.

For my setup I use a kubernetes config map so any change I make I need to reload the pods for now but for sure having the dashboards synchronized would be great!

**Comment by t0mer** [comment]
Well,
Finally got it working
AdGuardHome, 3 nodes clustering with Configuratrion Replication, Grafana Dashboard and even Uptime Monitor for each node (not included in the video).
https://youtu.be/zyz3jQ6_s4A

**Comment by tomlawesome** [comment]
> Well,
> Finally got it working
> AdGuardHome, 3 nodes clustering with Configuratrion Replication, Grafana Dashboard and even Uptime Monitor for each node (not included in the video).
> https://youtu.be/zyz3jQ6_s4A

Awesome. 

Can you please share config info? I've set up a new swarm, but am struggling to get 3 instances tied together and would really love to.

**Comment by t0mer** [comment]
> > Well,
> > Finally got it working
> > AdGuardHome, 3 nodes clustering with Configuratrion Replication, Grafana Dashboard and even Uptime Monitor for each node (not included in the video).
> > https://youtu.be/zyz3jQ6_s4A
> 
> Awesome.
> 
> Can you please share config info? I've set up a new swarm, but am struggling to get 3 instances tied together and would really love to.

Working on it, Hopefully will uploadto my git un a few days

**Comment by tomlawesome** [comment]
> > > Well,
> > > Finally got it working
> > > AdGuardHome, 3 nodes clustering with Configuratrion Replication, Grafana Dashboard and even Uptime Monitor for each node (not included in the video).
> > > https://youtu.be/zyz3jQ6_s4A
> > 
> > 
> > Awesome.
> > Can you please share config info? I've set up a new swarm, but am struggling to get 3 instances tied together and would really love to.
> 
> Working on it, Hopefully will uploadto my git un a few days

Thank you! Look forward to it, maybe I can resolve the setup for myself :)

**Comment by lensherm** [comment]
> > > Well,
> > > Finally got it working
> > > AdGuardHome, 3 nodes clustering with Configuratrion Replication, Grafana Dashboard and even Uptime Monitor for each node (not included in the video).
> > > https://youtu.be/zyz3jQ6_s4A
> > 
> > 
> > Awesome.
> > Can you please share config info? I've set up a new swarm, but am struggling to get 3 instances tied together and would really love to.
> 
> Working on it, Hopefully will uploadto my git un a few days

Thank you sir! I got a few Pi4s waiting to be configured as soon as you post your configs.

**Comment by lensherm** [comment]
@ameshkov , on a somewhat related note, I know this has been brought up a few times before, but is there any chance to bake the synchronization of the pertinent portion of the config into the product? Maybe split this subset of configuration parameters into a separate config doc? Never mind this if it's already on the roadmap. From the end user perspective, it just would be super nice to not have to set up elaborate load balancing, clustering, or virtual IP solutions for a home network.

In my specific case, I'm running AGH as a Home Assistant add-on, inside a VM. Every time I have to perform maintenance requiring restart of that VM, or the host machine, for example if there's an OS update, the DNS resolution in the entire house goes down. The reason I'm not serving another DNS server through DHCP, is because I'm using AGH to  limit what the kids get access to. It would be super nice to have a second instance of AGH running with similar configuration, somewhere in another VM, container, or even an RPi.

As an aside, since I started speaking about parental controls, is it possible to expand on the concept of tags to  arrive at  a semblance of client groups, OR to be able to filter the client list by using tags AND to perform bulk actions on the result of the filtering? Extra bonus if you could save these filtered lists, for later use. An example of the use case for this would be to turn off access to, let's say YouTube, during school time, not only on child devices, but also on the media players and smart TVs in the house.  And a double extra bonus would be if these operations could be performed through an API, so one could automate or schedule actions from external platforms, such as NodeRed, Homebridge, Home Assistant, etc.

Sorry for the long post and for stuffing more than one feature request in a single ticket.

**Comment by lensherm** [comment]
In the meantime, I'll try to do some rsync/udp load balance voodoo with something like this here:
https://hub.docker.com/r/instantlinux/udp-nginx-proxy

**Comment by ameshkov** [comment]
@lensherm well, what we should do prior to anything is implement the proper configuration reload that does not require restarting AGH. Only when we have this, implementing sync becomes doable.

**Comment by lensherm** [comment]
> @lensherm well, what we should do prior to anything is implement the proper configuration reload that does not require restarting AGH. Only when we have this, implementing sync becomes doable.

Sounds like a good plan, @ameshkov. One question on this, since the specific portions of the configuration we're discussing here are accessible through the UI and don't seem to require a restart, can't this subset of the configuration be kept in sync by issuing calls to the back ends of each required AGH instance, every time new settings are saved?

**Comment by ameshkov** [comment]
> can't this subset of the configuration be kept in sync by issuing calls to the back ends of each required AGH instance, every time new settings are saved?

Generally, one can use [AdGuard Home API](https://github.com/AdguardTeam/AdGuardHome/tree/master/openapi) and write a sync tool that does that. But this sort of solution would be quite hard to maintain.

**Comment by bakito** [comment]
Hi all

I'm currently working on a sync tool using the API as suggested by @ameshkov.

An early draft can be found here: https://github.com/bakito/adguardhome-sync

**Comment by Gorian** [comment]
Just to contribute to the discussion - adding multiple IPs to your DHCP config or `resolv.conf` isn't the only way of running multiple DNS servers - for the more technically inclined, I've had good luck with using BGP to provide anycast DNS locally which brings the benefit of HA to your DNS implementation - if one DNS server on your network goes down, BGP simply won't route to it anymore. I know it's out of scope for AdGuard to implement, but since people have been commenting about the "how to" of using multiple DNS servers and I didn't see anyone else mention anycast as a possible solution, I thought I'd mention it.

**Comment by tomlawesome** [comment]
> Just to contribute to the discussion - adding multiple IPs to your DHCP config or `resolv.conf` isn't the only way of running multiple DNS servers - for the more technically inclined, I've had good luck with using BGP to provide anycast DNS locally which brings the benefit of HA to your DNS implementation - if one DNS server on your network goes down, BGP simply won't route to it anymore. I know it's out of scope for AdGuard to implement, but since people have been commenting about the "how to" of using multiple DNS servers and I didn't see anyone else mention anycast as a possible solution, I thought I'd mention it.

It might not be in the scope but it’s definitely relevant (in my opinion) if existing technologies work perfectly well and serve the intended purpose. There’s no point in the AdGuard dev team reinventing the wheel. 

Thanks for the post, I’m off to do some reading about BGP/Anycast!

**Comment by Gorian** [comment]
No problem! I've been quite happy with anycast. I set this up with pihole (though I might check out AdGuard, lots of friends have been saying good things which is why I'm here checking it out :) )

I had issues where if the primary (first server listed in `resolv.conf`) was offline, then my clients would just not resolve dns, rather than using the second configure DNS server. It was really a pain. Rather than spend a lot of time troubleshooting *why* I found that anycast was just a better solution and side-stepped that whole problem. Plus I could do fun stuff like have a seperate IP range for BGP. Like LAN network might be `192.168.1.1/24`, with my DNS servers being `192.168.1.2`, and `.3`, but my BGP can be `10.0.0.53` for DNS to make it easy to remember

**Comment by Gorian** [comment]
> > can't this subset of the configuration be kept in sync by issuing calls to the back ends of each required AGH instance, every time new settings are saved?
> 
> Generally, one can use [AdGuard Home API](https://github.com/AdguardTeam/AdGuardHome/tree/master/openapi) and write a sync tool that does that. But this sort of solution would be quite hard to maintain.

It would be an architectural change, but if you moved to using sqlite and mysql for configuration instead of text files, it would be easier to maintain. You can default to use sqlite for single installations, with the option to use a mysql server/cluster if you have more than one. You can add support for redis if you want to sync user sessions. Then it wouldn't require as much maintenance beyond the initial implementation

**Comment by MZGSZM** [comment]
My use case for such a feature would mostly be used to reduce maintenance. I've got four different installs of AGH running currently and making sure they're all configured the same way and updated when I change one of them is kind of a hassle. I don't know if it would be possible for such a system to easily work over the internet as these installs are on separate internet connections (IE two on one, two on the other). Locally on the network only would be very helpful as well.

**Comment by jeremygaither** [comment]
@bakito To clarify, your plugin pushes the config (only) from a reference or primary node to multiple replica nodes? None of the logging is synchronized (probably doable elsewhere), correct?

**Comment by bakito** [comment]
@jeremygaither Correct, it copies the configuration. 
Logs and stats are not synched

**Comment by t0mer** [comment]
I have found a very simple way to copy the Setting & Cinfiguration from the main node to any other instance using [lsyncd](https://github.com/axkibe/lsyncd).

Here is a very simple script i wrote and it works great:
`settings {
logfile = "/var/log/lsyncd/lsyncd.log",
statusFile = "/var/log/lsyncd/lsyncd-status.log",
statusInterval = 3
}

-- Slave server configuration

sync {
default.rsync,
source="/opt/AdGuardHome",
exclude= {
        "data/*",
    },
target="192.168.0.131:/opt/AdGuardHome",
delay           = 0,

rsync = {
compress = true,
acls = true,
verbose = true,
owner = true,
group = true,
perms = true,
rsh = "/usr/bin/ssh -p 22 -o StrictHostKeyChecking=no"
}
}

sync {
default.rsync,
source="/opt/AdGuardHome",
target="192.168.0.132:/opt/AdGuardHome",
delay           = 0,
exclude= {
        "data/*",
    },
rsync = {
compress = true,
acls = true,
verbose = true,
owner = true,
group = true,
perms = true,
rsh = "/usr/bin/ssh -p 22 -o StrictHostKeyChecking=no"
}
}

sync {
default.rsync,
source="/opt/AdGuardHome",
target="/tmp",
delay           = 1,
exclude= {
        "data/*",
    },
rsync = {
binary = "/opt/AdGuardHome/adreload.sh",
compress = true,
acls = true,
verbose = true,
owner = true,
group = true,
perms = true,
rsh = "/usr/bin/ssh -p 22 -o StrictHostKeyChecking=no"
}
}
`

and after the sync process i'm sending restart command using this script:

`
#!/bin/bash
ssh root@192.168.0.131 "/opt/AdGuardHome/AdGuardHome -s restart"
ssh root@192.168.0.132 "/opt/AdGuardHome/AdGuardHome -s restart"
`

Hope you'll find it useful.

**Comment by t0mer** [comment]
> I'm not running AdGuard Home in k8s, although a clustered / redundant setup is something I'd like in future. However, I have been successfully ingesting the query log using Telegraf and injecting it into InfluxDB to be rendered by Grafana. I'm no Telegraf expert, but here's the config that's working for me:
> 
> ```toml
> [[inputs.tail]]
>   files = ["/opt/AdGuardHome/data/querylog.json"]
>   data_format = "json"
>   tag_keys = [
>     "IP",
>     "QH",
>     "QT",
>     "QC",
>     "CP",
>     "Upstream",
>     "Result_Reason"
>   ]
> 
>   json_name_key = "query"
>   json_time_key = "T"
>   json_time_format = "2006-01-02T15:04:05.999999999Z07:00"
> 
> [[processors.regex]]
>   [[processors.regex.tags]]
>     key = "IP"
>     result_key = "IP_24"
>     pattern = "^(\\d+)\\.(\\d+)\\.(\\d+)\\.(\\d+)$"
>     replacement = "${1}.${2}.${3}.x"
> 
> [[processors.regex]]
>   [[processors.regex.tags]]
>     key = "QH"
>     result_key = "TLD"
>     pattern = "^.*?(?P<tld>[^.]+\\.[^.]+)$"
>     replacement = "${tld}"
> 
> [[outputs.influxdb]]
>   urls = ["http://influxdb.local:8086"]
>   database = "adguard"
> ```
> 
> You'll see I've also included a few processors that give me some extra useful stats to chart, including the origin subnet (useful because I use VLANs which map to subnets) and the TLD of the requested domain. It would be easy to run a Telegraf agent in each AdGuard Home pod (mine is actually running in an LXC container on Proxmox rather than k8s) to centralize all your logging.
> 
> One thing to note is you'll also want to adjust the `querylog_size_memory` in the `AdGuardHome.yaml` file. This is how many log entries it keeps in memory before flushing to the query log. I think it defaults to 1,000 but I dropped it to 5 to keep the data flowing smoothly to Telegraf.
> 
> The end result looks like this:
> 
> <img alt="Screen Shot 2021-01-14 at 7 26 53 AM" width="1599" src="https://user-images.githubusercontent.com/726531/104591365-91414d80-563a-11eb-8239-b0b820bcfb85.png">
> 
> Hope that helps! @tomlawesome @onedr0p @mzac @jeremygaither

Works perfect, thanks for that.
![image](https://user-images.githubusercontent.com/4478920/120229567-1463e880-c256-11eb-9e53-273f1d1d37cf.png)

**Comment by teewhey** [comment]
Curious, is there any reason we don't want to consider using backend databases like MongoDB to store configs and stats?

**Comment by PSSGCSim** [comment]
MongoDB sounds like extreme overkill and resource hog. I'd prefer something simple like config sync over API (maybe one node with read-only settings) and stats/log shipping to destination of your choice.

I currently rsync my configs from "primary" node and ship logs with syslog to my centralized syslog server.

**Comment by onedr0p** [comment]
IMO storing configuration in an external distributed database (etcd, consul or whatever) isn't a terrible idea but it would require a huge application rewrite You would be able to scale to `n` number of pods without having to worry about which is a "primary" or "secondary" and just load balance between the pods.

**Comment by teewhey** [comment]
Well basically what we need is to host the config somewhere else.  Could consider using Redis even to store the config json string.

**Comment by zamnuts** [comment]
I really like the simplicity of adguard. A single binary and a config file. This makes it straight forward and intuitive to run in a container/k8s. Adding external data stores for managing configuration will increase complexity and runtime requirements, which IMO doesn't make sense for this project.

In a kubernetes environment, the only problem I've encountered is aggregating metrics among the replicaset. Configuration is straight-forward by means of a ConfigMap. If supporting a distributed datastore is desired, despite being an anti-pattern, I'd suggest something lightweight and simplistic like https://github.com/rqlite/rqlite, which could be embedded or internally bootstrapped (it is go!).

However, status and stats can be scraped via API (requiring authentication) and exported to prometheus. I haven't tried this, but seems promising. https://github.com/ebrianne/adguard-exporter

**Comment by Gorian** [comment]
as I stated before you can just *default* to using something like sqlite instead of a database, meaning that the default complexity won't increase, while allowing advanced users more options. Want simple? Great, don't change any configs. Want more complex? Great, change this config option to `redis` or `mongodb` or whatever, instead of sqlite. But it feels bad to limit users who want *more* flexibility and options, purely on the basis that some people won't use or understand them, especially when it's easy to just keep them hidden and unused unless a user specifically wants to use them.

**Comment by hoshsadiq** [comment]
Something else to consider if the option of clustering is added: how would DHCP be handled? I assume only a leader would be able to assign IPs.

**Comment by huiguang-liang** [comment]
+1. Would absolutely love to see some solution to this. I used to run AdGuard Home as a single docker instance on my unraid box, and all my home DNS traffic is forced through this instance. Obviously I learnt pretty quick that this means I have a single point of failure, so I set up another AdGuard instance on another physical server, so now I have a primary and secondary instance of AdGuard that is advertised on my network as the primary and secondary DNS server.

I don't see a transparent way to sync the configs that does not involve some host-level file copying. Having an option to use something like `redis` or `mongodb` seems like a perfect solution to this issue.

**Comment by huiguang-liang** [comment]
And additional note - I have about 20 containers running different services and many of them already require a `redis`, `postgres`, `mysql` and `mariadb` instance to persist data, so I have standalone containers for those databases. I suspect most folks will have similar setups, especially if you're deep into the self-hosting rabbit hole. So using some of these datastores to persist data helps to simplify this configuration sync without having to deal with any host-level file copying, permissions, and what not.

**Comment by Gorian** [comment]
> Something else to consider if the option of clustering is added: how would DHCP be handled? I assume only a leader would be able to assign IPs.

I'm confused what DHCP has to do with this? AFAIK, adguard is a DNS service, right? DNS shouldn't be service DHCP *as* well, that's an entirely separate service.

**Comment by zamnuts** [comment]
@Gorian AdGuard Home has a built-in DHCP server. In the web console, go to Settings > DHCP Settings. Also, see https://kb.adguard.com/en/home/overview#comparison-to-pi-hole which says:

Feature | AdGuard Home | Pi-Hole
-- | -- | --
Built-in DHCP server | ✅ | ✅

**Comment by Gorian** [comment]
huh. Okay, seems weird to me to bundle them.

**Comment by ainar-g** [comment]
Sorry for the late response, we've been busy with a lot of things lately.  It seems like most people just want a better way to synchronise configuration and persistent state between several AdGuardHome instances.  So our main goal is to eventually make the synchronisation work by simply copying `AdGuardHome.yaml` and the `data/` directory and calling `kill -HUP "$( pgrep AdGuardHome )"`.

That means that first of all we need to actually support full reload of the configuration file and the `data/` on `SIGHUP`, which is basically impossible to do in the current architecture, but is an explicit goal of the v0.108.0 release.  The query log synchronisation may be a bit tricker, but will probably be easier once we move to a better storage.

I'm not sure that BGP configuration and other ways of using several servers on a single IP address are in the scope of the project right now.  It's definitely not something we can bake into AGH in reasonable time.

I'll move this issue into the v0.108.0 milestone, since this is when `SIGHUP` handling and full configuration reload should arrive.

**Comment by scyto** [comment]
If i place the adguard data on a replicated FS like glusterfs and point two instances at the data will it work or corrupt the data?

**Comment by onedr0p** [comment]
@scyto no, that's not how that works

**Comment by siavashs** [comment]
I've created the following setup for redundancy:
- unbound as cache forwarding the `.` zone to AdGuardhome (Primary DNS server advertised by DHCP server)
- AdGuardHome (Secondary DNS server advertised by DHCP server)

Pros:
- Some level of redundancy (for planned maintenance or updates on either the cache or AdGuardHome)

Cons:
- From AdGuards point of view most requests are from the cache server
- If AdGuardHome is down, the cache can only respond to existing records (can be mitigated by allowing queries to another upstream if AdGuardHome is down)

**Comment by scyto** [comment]
> @scyto no, that's not how that works

to add details for others on why it doesn't work like that- one node can lock the state of the db file - which can cause the other node to hang on start at the `[info] Initializing auth module: /opt/adguardhome/work/data/sessions.db` stage.

**Comment by scyto** [comment]
> I've created the following setup for redundancy:

Approach i settled on is to run in a swarm with macvlan a single instance.  

So while this doesn't protect against process failure / process corruption it does protect against container failure, docker host failure, VM failure, VM host failure and physical hardware failure.

**Comment by Gorian** [comment]
> I've created the following setup for redundancy:
> 
> * unbound as cache forwarding the `.` zone to AdGuardhome (Primary DNS server advertised by DHCP server)
> * AdGuardHome (Secondary DNS server advertised by DHCP server)
> 
> Pros:
> 
> * Some level of redundancy (for planned maintenance or updates on either the cache or AdGuardHome)
> 
> Cons:
> 
> * From AdGuards point of view most requests are from the cache server
> * If AdGuardHome is down, the cache can only respond to existing records (can be mitigated by allowing queries to another upstream if AdGuardHome is down)

FYI, the reason I ran BGP at home is that in testing I found that a lot of servers didn't failover to the second configured DNS server if the first one is having issues - it just stop resolving altogether. Make sure you actually test the primary failing in different ways before relying on this.

**Comment by Dynasty-Dev** [comment]
Is it possible to add my second adguard home ip  in my first adguard home's instance dhcp settings? Would be great as my router doesnt allow me to change dns so i have to rely on adguards dhcp.

**Comment by emlimap** [comment]
I ended up using [entr](https://github.com/eradman/entr) to start a sync from Primary server to backup server whenever there is a change to config file. 

This means any changes to Primary instance is replicated to secondary within few seconds of config file being written to disk without relying on a scheduled cron.

entr is available in Debian & Ubuntu repos, so you should be able to install it by running `apt-get install entr`.

**Assumptions**
- You run AdGuardHome as a non root user, `aghome` in my case
- `aghome` user has key based SSH auth so you can login without password
- `aghome` user has password less sudo permission to stop & start AdGuardHome service
- User also has correct permissions to write config file to location AdGuardHome reads it from.
- AdGuardHome is installed `/opt/` path
- entr is installed on the primary server/server the config file should be synced from.

If the assumptions aren't true, feel free to change it to match your setup.

systemd service
```sh
[Unit]
Description=AdGuard Home sync service
After=AdGuardHome.service

[Service]
User=aghome
Group=aghome
StartLimitInterval=5
StartLimitBurst=10
ExecStart=/usr/local/scripts/sync-service.sh


Restart=on-failure
RestartSec=5

[Install]
WantedBy=multi-user.target
```

Shell script systemd calls. Couldn't use pipe in systemd service file so had to use a shell script for it. `-p` paramater ensures that it only calls the script once the file has been fully written to disk.

```sh
#!/usr/bin/env sh
/bin/echo /opt/AdGuardHome/AdGuardHome.yaml | /usr/bin/entr -p /usr/local/scripts/agh-sync.sh
```

`agh-sync.sh` Script to run when file changes

```sh
#!/usr/bin/env sh
echo "Starting AGH config sync"
ssh aghome@192.168.100.3 'sudo systemctl stop AdGuardHome'
rsync /opt/AdGuardHome/AdGuardHome.yaml aghome@192.168.100.3:/opt/AdGuardHome/AdGuardHome.yaml
ssh aghome@192.168.100.3 'sudo systemctl start AdGuardHome'
echo "Sync Complete"
```


In the backup instance, have the below in `/etc/sudoers` file so that user `aghome` can start & stop AdGuardHome without being prompted for password. Also these are the only sudo commands `aghome` can run and nothing else.

```sh
aghome ALL=(ALL) NOPASSWD: /bin/systemctl stop AdGuardHome, /bin/systemctl start AdGuardHome
```

Stopping & starting the service does take time and also clears the in memory cache adguardhome holds. So it would be nice at some point for AdGuard Home to support hot reloading config file without restarting the service.

**Comment by mazay** [comment]
Hi all, I've been running multiple AGH instances successfully for at least a year.

Config sync doesn't seem to be an issue, I'm just storing the `conf` dir on NFS share, I'm pretty sure any shared/replicated FS will do the trick.

The only issue I've had so far is databases as those are not meant to be used by multiple instances, so I have a fully working HA setup but somewhat broken stats. It would be great if that DB issue could be solved somehow but I can live with that.

**Comment by JaneJeon** [comment]
+1 to what the person above said.

Right now, it is _already possible_ to run multiple instances of AGH and make it HA by putting the instances behind a load balancer. The only thing is that the file-based database is meant to be written to by only one at a time, causing issues when multiple instances are writing to it (when you share the volumes between the two instances).

Therefore, the way to add HA to AGH is to simply allow usage of MySQL/Postgres as the database of choice; as everything else - config files and whatnot - will be effectively "stateless".

**Comment by kotx** [comment]
I think Adguard currently uses SQLite? The folks at @superfly made LiteFS which can replicate/distribute SQLite databases. It might be worth taking a look for deployments, however you'd either need to determine a hardcoded primary, setup Consul, or use Fly.io which automatically hosts Consul for you.

**Comment by JaneJeon** [comment]
...which is obviously a non-starter. I think it's _much_ easier to just extend whatever SQL layer they're using to support other SQL databases as well, which immediately solves this (and a whole host of other related issues) in one go.

**Comment by kotx** [comment]
> ...which is obviously a non-starter. 

Absolutely, there definitely should be a better solution.
But could be a workaround for end users right now, seeing as LiteFS uses FUSE, you could intercept any reads/writes to any sqlite db on the filesystem.

**Comment by scyto** [comment]
I have been running adguard home with [adguardhome sync](https://github.com/bakito/adguardhome-sync) for over a year now, has worked perfectly. It is realtively simple (i still agree a simpler native solution would rock, unfortunately i cannot code so i am of no use at all).

**Comment by celevra** [comment]
syncing is one thing, but we need stats from both (or more) nodes, so there is no way around mysql or any other sql database

**Comment by t0mer** [comment]
It can be done using Telegraf + InfluxDb + Grafana

![image](https://user-images.githubusercontent.com/4478920/228663536-9d35ef0f-657d-4c91-9a29-7fd462b55ec9.png)


> syncing is one thing, but we need stats from both (or more) nodes, so there is no way around mysql or any other sql database

**Comment by Hornochs** [comment]
> I'm not running AdGuard Home in k8s, although a clustered / redundant setup is something I'd like in future. However, I have been successfully ingesting the query log using Telegraf and injecting it into InfluxDB to be rendered by Grafana. I'm no Telegraf expert, but here's the config that's working for me:
> 
> ```toml
> [[inputs.tail]]
>   files = ["/opt/AdGuardHome/data/querylog.json"]
>   data_format = "json"
>   tag_keys = [
>     "IP",
>     "QH",
>     "QT",
>     "QC",
>     "CP",
>     "Upstream",
>     "Result_Reason"
>   ]
> 
>   json_name_key = "query"
>   json_time_key = "T"
>   json_time_format = "2006-01-02T15:04:05.999999999Z07:00"
> 
> [[processors.regex]]
>   [[processors.regex.tags]]
>     key = "IP"
>     result_key = "IP_24"
>     pattern = "^(\\d+)\\.(\\d+)\\.(\\d+)\\.(\\d+)$"
>     replacement = "${1}.${2}.${3}.x"
> 
> [[processors.regex]]
>   [[processors.regex.tags]]
>     key = "QH"
>     result_key = "TLD"
>     pattern = "^.*?(?P<tld>[^.]+\\.[^.]+)$"
>     replacement = "${tld}"
> 
> [[outputs.influxdb]]
>   urls = ["http://influxdb.local:8086"]
>   database = "adguard"
> ```
> 
> You'll see I've also included a few processors that give me some extra useful stats to chart, including the origin subnet (useful because I use VLANs which map to subnets) and the TLD of the requested domain. It would be easy to run a Telegraf agent in each AdGuard Home pod (mine is actually running in an LXC container on Proxmox rather than k8s) to centralize all your logging.
> 
> One thing to note is you'll also want to adjust the `querylog_size_memory` in the `AdGuardHome.yaml` file. This is how many log entries it keeps in memory before flushing to the query log. I think it defaults to 1,000 but I dropped it to 5 to keep the data flowing smoothly to Telegraf.
> 
> The end result looks like this:
> 
> <img alt="Screen Shot 2021-01-14 at 7 26 53 AM" width="1599" src="https://user-images.githubusercontent.com/726531/104591365-91414d80-563a-11eb-8239-b0b820bcfb85.png">
> 
> Hope that helps! @tomlawesome @onedr0p @mzac @jeremygaither

Hey there,
I tried to build this yesterday, but there are several stones in my way. The Dashboard JSON isn't working anymore cause the Piechart Plugin doesn'T exist anymore for example. Do you got some updated Configs and Dashbord maybe? Thanks!

**Comment by drabgail** [comment]
Jumping in on this. Grafana is good for combining the stats but I think what would be useful is to have the ability to just see those stats from your multiple AG instances in the query log of anyone one instance, you would still have the ability to perform contextual actions on those queries. Right now I have two instances running with the inotify/rsync method on the config file. If I have an item I want to block and I go to my AG query log it's 50/50 as to whether I see it there. I either need to block that item manually or wait for it to be queried by the primary AG, else the config change will just be lost when the primary is updated.

**Comment by ilMagnifico** [comment]
Bumping this. It's very inconvenient to have to apply the same custom rules across different devices. They should sync to each other.

**Comment by BCZ0101** [comment]
> I'm not running AdGuard Home in k8s, although a clustered / redundant setup is something I'd like in future. However, I have been successfully ingesting the query log using Telegraf and injecting it into InfluxDB to be rendered by Grafana. I'm no Telegraf expert, but here's the config that's working for me:
> 
> ```toml
> [[inputs.tail]]
>   files = ["/opt/AdGuardHome/data/querylog.json"]
>   data_format = "json"
>   tag_keys = [
>     "IP",
>     "QH",
>     "QT",
>     "QC",
>     "CP",
>     "Upstream",
>     "Result_Reason"
>   ]
> 
>   json_name_key = "query"
>   json_time_key = "T"
>   json_time_format = "2006-01-02T15:04:05.999999999Z07:00"
> 
> [[processors.regex]]
>   [[processors.regex.tags]]
>     key = "IP"
>     result_key = "IP_24"
>     pattern = "^(\\d+)\\.(\\d+)\\.(\\d+)\\.(\\d+)$"
>     replacement = "${1}.${2}.${3}.x"
> 
> [[processors.regex]]
>   [[processors.regex.tags]]
>     key = "QH"
>     result_key = "TLD"
>     pattern = "^.*?(?P<tld>[^.]+\\.[^.]+)$"
>     replacement = "${tld}"
> 
> [[outputs.influxdb]]
>   urls = ["http://influxdb.local:8086"]
>   database = "adguard"
> ```
> 
> You'll see I've also included a few processors that give me some extra useful stats to chart, including the origin subnet (useful because I use VLANs which map to subnets) and the TLD of the requested domain. It would be easy to run a Telegraf agent in each AdGuard Home pod (mine is actually running in an LXC container on Proxmox rather than k8s) to centralize all your logging.
> 
> One thing to note is you'll also want to adjust the `querylog_size_memory` in the `AdGuardHome.yaml` file. This is how many log entries it keeps in memory before flushing to the query log. I think it defaults to 1,000 but I dropped it to 5 to keep the data flowing smoothly to Telegraf.
> 
> The end result looks like this:
> 
> <img alt="Screen Shot 2021-01-14 at 7 26 53 AM" width="1599" src="https://user-images.githubusercontent.com/726531/104591365-91414d80-563a-11eb-8239-b0b820bcfb85.png">
> 
> Hope that helps! @tomlawesome @onedr0p @mzac @jeremygaither

Hello,
I have successfully configured telegraf, data records are also displayed in InfluxDB (_measurement > tail)...

I also successfully imported the dashboard. What I'm having trouble with is getting the dashboard to work and it always shows me:
Datasource AdGuard was not found.
Can someone help me here? In Grafna I see the AdGuard (telegraf) database.

**Comment by BCZ0101** [comment]
Hi
i got this Error:
A
Status: 500. Message: invalid: compilation failed: error @1:14-1:23: string literal key Elapsed must have a value error @1:37-1:76: expected comma in property list, got ASSIGN error @1:44-1:75: string literal key Upstream must have a value error @1:55-1:75: invalid expression @1:74-1:75: ' error @1:55-1:75: unexpected token for property key: ASSIGN (=) error @1:81-1:82: invalid statement: $ error @1:107-1:110: unexpected token for property key: DURATION (15m) error @1:111-1:112: invalid statement: ,

![1](https://github.com/user-attachments/assets/e4830443-88aa-4dfd-85c3-0f18c792fde6)

**Comment by Cansybars** [comment]
I didn't read the whole thread, but I'm planning this myself. Why not just have a single yaml file and instead of external IP for both listening and UI use 0.0.0.0. No synchronizatioon, most likely the first one will be hit and if not the second one either way you're making a change to the same file for both what is the problem with that? You just need to have a shared drive and file (e.g. External drive connected to AdGuard Home on raspberry pie and another instance on Windows), so completely done and no dependencies

**Comment by drabgail** [comment]
> I didn't read the whole thread, but I'm planning this myself. Why not just have a single yaml file and instead of external IP for both listening and UI use 0.0.0.0. No synchronizatioon, most likely the first one will be hit and if not the second one either way you're making a change to the same file for both what is the problem with that? You just need to have a shared drive and file (e.g. External drive connected to AdGuard Home on raspberry pie and another instance on Windows), so completely done and no dependencies

You could do this with a single yaml but unless you start making it more complex with overlays, etc. you will lose the redundancy as once your yaml file is offline all your instances are offline. I don't know for sure but I imagine that committing a change to that YAML on one instance would invoke a procedure to reload it but your other instances would not have this trigger so would not reload. 

In the end I went with rsync with a trigger on the yaml file so it pushes the changes to other instances and triggers the reload. I completely dropped grafana as it was too much of a pain for very little benefit, I just set my dns priority for specific services so I (almost) all the time know which instance to look at to see the dns queries for a given client.

**Comment by basalisco** [comment]
> I'm not running AdGuard Home in k8s, although a clustered / redundant setup is something I'd like in future. However, I have been successfully ingesting the query log using Telegraf and injecting it into InfluxDB to be rendered by Grafana. I'm no Telegraf expert, but here's the config that's working for me:
> 
> [[inputs.tail]]
>   files = ["/opt/AdGuardHome/data/querylog.json"]
>   data_format = "json"
>   tag_keys = [
>     "IP",
>     "QH",
>     "QT",
>     "QC",
>     "CP",
>     "Upstream",
>     "Result_Reason"
>   ]
> 
>   json_name_key = "query"
>   json_time_key = "T"
>   json_time_format = "2006-01-02T15:04:05.999999999Z07:00"
> 
> [[processors.regex]]
>   [[processors.regex.tags]]
>     key = "IP"
>     result_key = "IP_24"
>     pattern = "^(\\d+)\\.(\\d+)\\.(\\d+)\\.(\\d+)$"
>     replacement = "${1}.${2}.${3}.x"
> 
> [[processors.regex]]
>   [[processors.regex.tags]]
>     key = "QH"
>     result_key = "TLD"
>     pattern = "^.*?(?P<tld>[^.]+\\.[^.]+)$"
>     replacement = "${tld}"
> 
> [[outputs.influxdb]]
>   urls = ["http://influxdb.local:8086"]
>   database = "adguard"
> You'll see I've also included a few processors that give me some extra useful stats to chart, including the origin subnet (useful because I use VLANs which map to subnets) and the TLD of the requested domain. It would be easy to run a Telegraf agent in each AdGuard Home pod (mine is actually running in an LXC container on Proxmox rather than k8s) to centralize all your logging.
> 
> One thing to note is you'll also want to adjust the `querylog_size_memory` in the `AdGuardHome.yaml` file. This is how many log entries it keeps in memory before flushing to the query log. I think it defaults to 1,000 but I dropped it to 5 to keep the data flowing smoothly to Telegraf.
> 
> The end result looks like this:
> 
> <img alt="Screen Shot 2021-01-14 at 7 26 53 AM" width="1599" src="https://user-images.githubusercontent.com/726531/104591365-91414d80-563a-11eb-8239-b0b820bcfb85.png">
> 
> Hope that helps! [@tomlawesome](https://github.com/tomlawesome) [@onedr0p](https://github.com/onedr0p) [@mzac](https://github.com/mzac) [@jeremygaither](https://github.com/jeremygaither)

not read data DB is blank

```
root@adguard:/# telegraf -debug
2025-05-17T05:57:51Z I! Loading config: /etc/telegraf/telegraf.conf
2025-05-17T05:57:51Z I! Starting Telegraf 1.34.3 brought to you by InfluxData the makers of InfluxDB
2025-05-17T05:57:51Z I! Available plugins: 239 inputs, 9 aggregators, 33 processors, 26 parsers, 63 outputs, 6 secret-stores
2025-05-17T05:57:51Z I! Loaded inputs: tail
2025-05-17T05:57:51Z I! Loaded aggregators:
2025-05-17T05:57:51Z I! Loaded processors: regex (2x)
2025-05-17T05:57:51Z I! Loaded secretstores:
2025-05-17T05:57:51Z I! Loaded outputs: influxdb
2025-05-17T05:57:51Z I! Tags enabled: host=adguard
2025-05-17T05:57:51Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"adguard", Flush Interval:10s
2025-05-17T05:57:51Z W! [agent] The default value of 'skip_processors_after_aggregators' will change to 'true' with Telegraf v1.40.0! If you need the current default behavior, please explicitly set the option to 'false'!
2025-05-17T05:57:51Z D! [agent] Initializing plugins
2025-05-17T05:57:51Z D! [processors.regex] tags: Using explicit mode...
2025-05-17T05:57:51Z D! [processors.regex] tags: Using explicit mode...
2025-05-17T05:57:51Z D! [processors.regex] tags: Using explicit mode...
2025-05-17T05:57:51Z D! [processors.regex] tags: Using explicit mode...
2025-05-17T05:57:51Z D! [agent] Connecting outputs
2025-05-17T05:57:51Z D! [agent] Attempting connection to [outputs.influxdb]
2025-05-17T05:57:51Z D! [agent] Successfully connected to outputs.influxdb
2025-05-17T05:57:51Z D! [agent] Starting service inputs
2025-05-17T05:57:51Z D! [inputs.tail] Tail added for "/opt/AdGuardHome/data/querylog.json"
2025-05-17T05:58:01Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:58:11Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:58:21Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:58:31Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:58:41Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:58:51Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:59:01Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:59:11Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:59:21Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:59:31Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
2025-05-17T05:59:41Z D! [outputs.influxdb] Buffer fullness: 0 / 10000 metrics
```

**Comment by get-loose** [comment]
Just FYI, this is my simple approach using the adguardhome-sync container

In adguardhome.yaml you can set this 

``` yaml
dhcp:
  enabled: true
  icmp_timeout_msec: 1000
  options:
    - 6 ips 10.0.0.12,10.0.0.13,10.0.0.14
    - 15 text wub.search.wub.wub
    - 42 ips 10.0.0.135
```

The part that matters : setting option 6. It sets the list of dns servers that dhcp will tell it's clients about.
[Here](https://github.com/AdguardTeam/AdGuardHome/wiki/DHCP) and [here](https://datatracker.ietf.org/doc/html/rfc2132).

If we are assuming you have several AdguardHome on your network, hopefully one of them allows you to directly edit the yaml as described.  Then that one can be synced to the others (like HomeAssistantOS, no easy access to the yaml) using adguardhome-sync.  I disabled syncing the dhcp server settings (it's an option you can set in adguardhome-sync) from my main to the clones, since I only want 1 dhcp server on my lan.

Static leases just need to be set on main, Adguardhome automatically takes care of adding all DHCP leases to the DNS rewrites. That way they can be resolved by dns. And these do get synced to the clones (not that it matters much).

Last bit of oddity about the syncing : you can only sync dhcp leases (or static) to the clones if you enter dhcp-settings on them, and fill out everything as if you are activating a dhcp-server. So set

```  yaml
    network interface: eth9000
    gateway_ip: 10.0.0.135
    subnet_mask: 255.255.255.0
    range_start: 10.0.0.190
    range_end: 10.0.0.199
    lease_duration: 86400
```
on the clone and save but don't enable, then restart.  Only then can syncing succeed for the leases.

I'm still testing, only did this today, maybe I missed something.

GUI-solution would be nice though. But I like what I have now, no worries.

**Comment by zerwes** [comment]
hello @get-loose 
thx for sharing your experience with the dhcp options

> The word "option" at same indentation as icmp, the "dashes" two spaces to the right.

the next time you can use [(github) markdown for code formatting](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks)
```yaml

dhcp:
  enabled: true
  icmp_timeout_msec: 1000
  options:
    - 6 ips 10.0.0.12,10.0.0.13,10.0.0.14
    - 15 text wub.search.wub.wub
    - 42 ips 10.0.0.135
```
this makes it much more readable and frees you from the need for complicated explanations ...

**Comment by get-loose** [comment]
> this makes it much more readable and frees you from the need for complicated explanations ...

I was using the "code button <>",  but was placing my code between the two ticks of the first quote, not between the first and the second. And I tried multiple times lol. Hayaaa

**Comment by Thirawyn91** [comment]
Hello everyone,

I think option 6 is a good solution for setting up redundancy on both servers, but I've encountered a bug: it only works if the server's IP address is in the first position, otherwise Adguard removes it and responds to DHCP clients with the other servers in the option.

So I can't force one of the two servers to be the primary DNS.

If you have a solution ;)

**Comment by tenfourty** [comment]
I run two LXC containers with AdGuard Home (e.g. on IPs .253/.254), and then I use keepalived and a VIP (e.g. .252) for high availability, if one goes down, the other takes the VIP. I do this on every VLAN in my homelab.

**Comment by Thirawyn91** [comment]
Hi @tenfourty 

Thanks, I configure keepalived and it's works fine.
